rpc: Add importmempool RPC

maflcko opened the pull request:
Currently it is possible to import a mempool by placing it in the datadir and starting the node. However this has many issues:

* Users aren't expected to fiddle with the datadir, possibly corrupting it
* An existing mempool file in the datadir may be overwritten
* The node needs to be restarted
* Importing an untrusted file this way is dangerous, because it can corrupt the mempool

Fix all issues by adding a new RPC.

Commit:
doc: Clarify the getmempoolinfo.loaded RPC field documentation

Also, clarify the LoadMempool doxygen.

Commit:
refactor: Add and use kernel::ImportMempoolOptions

This allows optional named arguments with default values.

Commit:
Add importmempool RPC

test_importmempool_union contributed by glozow

Co-authored-by: glozow <gloriajzhao@gmail.com>

1440000bytes commented:
Hi @MarcoFalke 

More than 5 tests are failing

kevkevinpal commented on src/rpc/mempool.cpp:
```diff
@@ -719,6 +719,56 @@ static RPCHelpMan getmempoolinfo()
+             RPCArgOptions{.oneline_description = "\"options\""}},
+        },
+        RPCResult{RPCResult::Type::OBJ, "", "", std::vector<RPCResult>{}},
+        RPCExamples{HelpExampleCli("importmempool", "") + HelpExampleRpc("importmempool", "")},
```
would it make sense to add a `./pathto/mempool.dat` to the example param because right now when I type
`bitcoin-cli importmempool` 

I get

```
Examples:
> bitcoin-cli importmempool
> curl --user myusername --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "importmempool", "params": []}' -H 'content-type: text/plain;' http://127.0.0.1:8332/
```

This might be confusing
```suggestion
        RPCExamples{HelpExampleCli("importmempool", "./pathto/mempool.dat") + HelpExampleRpc("importmempool", "./pathto/mempoo.dat")},
```

ajtowns commented on src/kernel/mempool_persist.h:
```diff
@@ -12,15 +12,18 @@ class CTxMemPool;
+struct ImportMempoolOptions {
+    fsbridge::FopenFn mockable_fopen_function{fsbridge::fopen};
+    bool apply_fee_delta_priority{true};
+    bool apply_unbroadcast_set{true};
```
Should this also include the ability to override the (added to the mempool) `nTime` data with the current time (or an explicitly nominated time)?

Having a child transaction with an earlier `nTime` than its parent transaction might be confusing. I think the only way that can currently happen is with system/mock time shenanigans.

ajtowns commented on src/rpc/mempool.cpp:
```diff
@@ -719,6 +719,56 @@ static RPCHelpMan getmempoolinfo()
+{
+    return RPCHelpMan{
+        "importmempool",
+        "Import the file and attempt to add its contents to the mempool.\n"
```
Change "the file" to "a mempool.dat file" ?

Maybe reference the "savemempool" rpc as a way of updating mempool.dat with the current mempool?

ajtowns replied:
(I more just meant "what the heck type of file do you want me to give it???")

ajtowns reviewed (commented):
Approach ACK

ajtowns commented on src/rpc/mempool.cpp:
```diff
@@ -719,6 +719,56 @@ static RPCHelpMan getmempoolinfo()
+                 {"apply_unbroadcast_set", RPCArg::Type::BOOL, RPCArg::Default{false},
+                  "Whether to apply the unbroadcast set metadata from the mempool file."
+                  "Warning: Importing untrusted metadata may corrupt the "
+                  "mempool state."},
```
"undesirable behaviour" instead of bugs -- this is a case of you telling it to do the wrong thing, so not really a bug? Otherwise sgtm.

instagibbs commented:
concept ACK, will review

instagibbs commented on test/functional/mempool_persist.py:
```diff
@@ -159,6 +159,15 @@ def run_test(self):
+        fees = self.nodes[0].getmempoolentry(txid=last_txid)["fees"]
+        assert_equal(fees["base"], fees["modified"])
+        assert_equal({}, self.nodes[0].importmempool(mempooldat0, {"apply_fee_delta_priority": True, "apply_unbroadcast_set": True}))
+        fees = self.nodes[0].getmempoolentry(txid=last_txid)["fees"]
```
I'd like the test to load against two mempool files with differing txn subsets at least, showing that "additional" txs are loaded even if they weren't there originally. Even better if they're not strict subsets.

glozow replied:
Wrote a test, feel free to take / not take: https://github.com/glozow/bitcoin/commit/893f79c5ff612122e5b98d65b9859e7260c51a95#diff-99a8c43ebc6b8546e5fb9bf74032e96a6d3376b65655837e12454425098b9ad3R222

glozow commented on src/kernel/mempool_persist.cpp:
```diff
@@ -73,7 +72,7 @@ bool LoadMempool(CTxMemPool& pool, const fs::path& load_path, Chainstate& active
             CAmount amountdelta = nFeeDelta;
-            if (amountdelta) {
+            if (amountdelta && opts.apply_fee_delta_priority) {
                 pool.PrioritiseTransaction(tx->GetHash(), amountdelta);
```
`PrioritiseTransaction` applies the delta on top, so if txmempool already has a tx prioritised + this mempool.dat contains priority for it, this will stack them instead of rewriting the existing priority. I don't know if this is problematic, but probably good to keep in mind? (Wrote a test)

glozow replied:
Sorry for being pedantic, yes please. Something like "fee delta priority from the file is added on top of any priority already present in the mempool"

glozow reviewed (commented):
Concept ACK

svanstaa commented:
Why is `use_current_time` defaulting to true? As far as i understand, this would set all the broadcast times of the transactions in mempool.dat to the current time, effectively extending their lifetimes, to the point that (time-) expired txns would be dumped into the network again.
So I would suggest defaulting the value to false, unless there is some rationale behind it that is not obvious.

svanstaa commented:
Concept ACK

kristapsk commented:
Concept ACK

glozow reviewed (commented):
ACK fa83be1c30d5b181c024a40bfb9d5ea15ce64aec

mzumsande reviewed (commented):
Should the rpc be executable before the mempool load from `ThreadImport` has finished? I tried importing another mempool before the mempool from the datadir had finished loading, and I didn't encounter any problems on first sight - but I'm not sure how useful that would be and I guess the final result would be dependent on races between the two mempools being loaded in parallel?

Commit:
Remove Chainstate::LoadMempool

The 3-line function is only called once outside of tests, so it is
clearer to inline it.

ajtowns commented:
utACK fa776e61cd64a5ffd9a4be589ab8efeb5421861a

glozow reviewed (commented):
reACK fa776e61cd64a5ffd9a4be589ab8efeb5421861a

achow101 commented:
ACK fa776e61cd64a5ffd9a4be589ab8efeb5421861a

fanquake commented:
Release notes added in #28637.
